### ISSUE TITLE
Updating the linter

### ISSUE THREAD

**Issue by Jolanrensen** [Kotlin/dataframe#364]
The linter plugin we use (https://github.com/jeremymailen/kotlinter-gradle) can be updated to v3.14.0.

However, two things:

 - The excluding of rules is now handled by the .editorconfig file and this seems to be broken atm: https://github.com/jeremymailen/kotlinter-gradle/issues/324
 - Currently only changed files are linted, but the project is linted relatively poorly, so adding an updated linter would require a complete linting of every file in the project.
 - We need to check if we still get unused imports for extension functions referenced in KDocs.

**Comment by Jolanrensen** [comment]
After that PR is merged, the linter can be updated relatively easily.

- We need to wait till there are not so many PRs anymore
- The linter will format the entire project (!), most things can be done automatically, but the rest can be done in about a day's work, as seen in this example branch: https://github.com/Kotlin/dataframe/tree/k2-linter-attempt
- This branch also contains a sample .editorconfig file which disables only the rule "standard:filename" and it configures IntelliJ to not have *-imports at all :)
- Generated code can be excluded
- Using the [IntelliJ KtLint plugin](https://pinterest.github.io/ktlint/latest/install/setup/) we can more easily see errors and `@Suppress()` rules if we need to deviate somewhere.
- We can commit-hook the formatKotlin task so only clean code gets pushed :)
- We can even write custom rules if we need them (I don't expect we do) https://pinterest.github.io/ktlint/latest/api/custom-rule-set/

**Comment by Jolanrensen** [comment]
I've been tweaking some settings in .editorconfig to what I think looks best (and most in-line with DataFrame):
https://github.com/Kotlin/dataframe/blob/k2-linter-attempt/.editorconfig

We just need to figure out how to NOT run on generated-sources by default (as it's way to heavy) but allow the linter to format generated-sources from github actions instead.

**Comment by Jolanrensen** [comment]
Some updates:

I've had difficulties using Kotlinter due to its limited customization options. After trying Spotless, which also wasn't fit for the job, I found https://github.com/JLLeitschuh/ktlint-gradle, which is also used by kotlin-jupyer.

This branch contains a working project with that plugin: https://github.com/Kotlin/dataframe/tree/k2-ktlint-jlleitschuh-attempt.

In principle, it works the same as Kotlinter; it takes the .editorconfig file as source of truth, but, if we need it, we can programmatically override certain rules. It's ktlint-version independent. It can auto-check and -format when needed and it creates tasks for each individual source set. Since the generated-sources are no sourceSet, it will not run on them by default, which is great :) We can still create a custom ktlint format task for the generated sources though, as I did in `:core:ktlintFormatGeneratedSources` in the branch. This can run automatically after `processKDocsMain` and format the generated-sources folder :) (with `outputReadOnly = false` of couse).
